data/blog/how-to-refactor-and-not-break-things.md

```diff
@@ -0,0 +1,101 @@
+
+Again, our test suite gave us confidence that the SDK still behaved the same as before.
+
+## Phase III: Use New Scopes Everywhere
+
+In the Python SDK, we have over 40 integrations for various web frameworks, databases, and other libraries. All of them still used the Hub-based API.
+
+We updated each integration from the old Hub-based API to the new Scope-based API in a separate PR. This was pretty straightforward and done in a couple of days.
+
+Doing this gave us insights into the look and feel of the new API. Because we now used the API in the same way as our power users would use it in the future. We found some things that we did not like and made some minor changes to the new API to make it more convenient to use. Having one PR for each integration created a lot of small PRs that were easy to review.
+
+After we had all PRs merged, we created a release candidate and released it on PyPI so everyone could give it a try and we could gather feedback.
+
+We updated our internal usage dashboards so we could see the adoption of the new release candidate.
+
+## Phase IV: Load Testing and Dogfooding
+
+We started to do load testing. We created a [sample application and ran load tests](https://github.com/getsentry/demo-flask-load-test) locally to check if the CPU and memory usage would change between the current SDK and the new 2.0 version. It did not. After the load tests, we were confident that we could use the release candidate on Sentry.io.
+
+Sentry.io is a very big Python project where we use all the advanced features of the SDK. It is a perfect candidate for dogfooding because it will uncover problems very soon.
+
+We first installed the new SDK on our Canary servers. We let it run for an hour and closely monitored CPU and Memory. Everything looked good. Like with the local load tests, nothing spiked.
+
+Time for some proper dogfooding.
+
+We installed SDK 2.0 on **all** our servers. After running the new version for a week on Sentry.io we discovered that there was a problem in the Celery integration where the baggage header (used for propagating trace information) was handled incorrectly. After fixing this we created a new release candidate and continued the dogfooding.
+
+All in all, we created 6 release candidates where we each time fixed some smaller problems or made improvements.
+
+## Finally: The Release
+
+When we were confident that all problems were fixed, all data was correct, and CPU and memory usage was the same as before we [released version 2.0](https://github.com/getsentry/sentry-python/releases/tag/2.0.0).
+
+On the first day, there was a [bug report](https://github.com/getsentry/sentry-python/issues/3021) from a user when using the SDK with the Starlette framework and Uvicorn as the server. We fixed the problem and released [2.0.1](https://github.com/getsentry/sentry-python/releases/tag/2.0.1) right away.
```

Comment:
Oh OK so yeah it's better to reword the paragraph above indeed, since a patch was needed immediately.